### PR TITLE
added corpus_iter for Terrier index

### PR DESCRIPTION
**Pull request by cmacdonald**
Initial draft of https://github.com/terrier-org/pyterrier/issues/425

Unit tests are needed. 

@seanmacavaney is the API useful?

### CODE REVIEW

File: pyterrier/bootstrap.py [hunk 1]
```diff
@@ -271,14 +271,55 @@ def _index_add(self, other):
...
+            keys_offset = {'docno' : keys_offset['docno'] }
+            direct_inputstream = self.getIndexStructureInputStream("direct")
+            lex = self.getLexicon()
+
+            while direct_inputstream.hasNext():
+                ip = direct_inputstream.getNextPostings() #this is the next() method
+
+                for skipped in range(0, direct_inputstream.getEntriesSkipped()):
+                    meta = meta_inputstream.next()
+                    yield {k : meta[keys_offset[k]] for k in keys_offset}
```
Comment (seanmacavaney) [inline pyterrier/bootstrap.py:297]: Not super familar with the semantics of `getEntriesSkipped`. But should this include an empty `toks` to ensure that every dict from the iter has the same fields?

File: pyterrier/bootstrap.py [hunk 2]
```diff
@@ -271,14 +271,55 @@ def _index_add(self, other):
             raise ValueError("Cannot document-wise merge indices with and without positions (%r vs %r)" % (blocks_1, blocks_2))
         multiindex_cls = autoclass("org.terrier.realtime.multi.MultiIndex")
         return multiindex_cls([self, other], blocks_1, fields_1 > 0)
+    
+    def _index_corpusiter(self, direct=True):
```
Comment (seanmacavaney) [inline pyterrier/bootstrap.py:275]: `direct` isn't the most intuitive name to me for this field. Perhaps `return_toks`?

File: docs/terrier-index-api.rst [hunk 3]
```diff
@@ -171,3 +171,31 @@ of the term (obtained from the Lexicon, in the form of the LexiconEntry), as wel
...
+
+    iter = index.get_corpus_iter(direct=False)
+    next(iter)
+    ## could display {'docno' : 'd1', 'text' : 'This document contains ...' }
+
+Pre-tokenised Example::
+
+    iter = index.get_corpus_iter()
+    next(iter)
+    ## could display {'docno' : 'd1', 'toks' : {'a' : 1, 'the' : 2}}
```
Comment (seanmacavaney) [inline docs/terrier-index-api.rst:192]: should also include "text"?

**Review by seanmacavaney — COMMENTED** [review]
looks good, just a few minor suggestions

File: pyterrier/bootstrap.py (hunk 1 above)
Comment (cmacdonald) [reply pyterrier/bootstrap.py:297]: it should, yes

File: docs/terrier-index-api.rst (hunk 3 above)
Comment (cmacdonald) [reply docs/terrier-index-api.rst:192]: if text is present in the meta index

File: pyterrier/bootstrap.py (hunk 2 above)
Comment (cmacdonald) [reply pyterrier/bootstrap.py:275]: or `toks`? The corresponding indexing arg is `pretokenised`?

File: pyterrier/bootstrap.py (hunk 1 above)
Comment (cmacdonald) [reply pyterrier/bootstrap.py:297]: > Not super familar with the semantics of getEntriesSkipped.

direct_inputstream only address documents that are non-empty. So we have to address non-empty documents using getEntriesSkipped()

**Comment by cmacdonald** [comment]
> direct isn't the most intuitive name to me for this field.

Final decision: `toks` or `return_toks` ?

**Comment by seanmacavaney** [comment]
I don't think I feel super strongly one way or the other, but I feel `return_toks` is a bit clearer (i.e., it sounds like a boolean).

**Comment by cmacdonald** [comment]
Ok, ta. I write some unit tests then I merge.

**Comment by cmacdonald** [comment]
My revised implementation requires Python 3.8.
Python 3.7 is EOL (July 2023). Should enforce an upgrade?

**Comment by seanmacavaney** [comment]
I'm happy with min python version of 3.8

**Pull request merged**